### ISSUE TITLE
Question regarding inputs/outputs of orbit.Orbit.from_fit() 

### ISSUE THREAD

**Issue by callous4567** [jobovy/galpy#467]
Hello;

I've a question regarding making effective use of the orbit.Orbit.from_fit() feature in Galpy, specifically with regard to the input parameters and interpreting the output data of orbit.Orbit.getOrbit(). 

Presently, I am setting up the input vxvv/from_fit, via (where all inputs are in the units requested by the Wiki- deg deg kpc mas/yr mas/yr km/s) 

```
# Set up the data array 
vxvv = np.empty(shape=(len(data_to_fit), 6))
vxvv[:,0], vxvv[:,1], vxvv[:,2], vxvv[:,3],vxvv[:,4], vxvv[:,5] = data_to_fit['l'],\
                                                                                        data_to_fit['b'],\
                                                                                        data_to_fit['dist'],\
                                                                                        data_to_fit['dmu_l']*np.cos(np.radians(data_to_fit['b'])),\
                                                                                        data_to_fit['dmu_b'],\
                                                                                        data_to_fit['vlos']

# Also set up the errors. Don't forget: edmu is multiplied by cos(dec) in Galpy.
evxvv = np.empty(shape=(len(data_to_fit), 6))
evxvv[:,0], evxvv[:,1], evxvv[:,2], evxvv[:,3],evxvv[:,4], evxvv[:,5] = np.zeros(len(data_to_fit)),\
                                                                        np.zeros(len(data_to_fit)),\
                                                                        data_to_fit['edist'],\
                                                                        data_to_fit['edmu_l']*np.cos(np.radians(data_to_fit['b'])),\
                                                                        data_to_fit['edmu_b'],\
                                                                        data_to_fit['evlost']

# Set up the initial orbit guess
init_vxvv = np.mean(vxvv, axis=0)

# Set up the fit for the data.
fit = galpy.orbit.Orbit.from_fit(init_vxvv,
                                                 vxvv,
                                                 evxvv,
                                                 orbigist.pot,
                                                 ro=orbigist.rovo[0]*u.kpc,
                                                 vo=orbigist.rovo[1]*u.km/u.s,
                                                 zo=orbigist.zo*u.kpc,
                                                 solarmotion=solarmotion,
                                                 lb=True)
```

where the potential has the same ro,vo,etc/etc as in this case (Hern bulge, MN disk, NFW halo.) Once the fit has been produced, I integrate it and extract all the various coordinates (in this example using bovy_coords) via

```
# Get a "flip" for backward orbit
fit_backward = fit.flip()

# Get integrals
fit.integrate(np.linspace(0, 0.1e9, 1000)*u.yr, orbigist.pot, 'rk4_c')
fit_backward.integrate(np.linspace(0, 0.1e9, 1000)*u.yr, orbigist.pot, 'rk4_c')

# Get fit coordinates in galcencyl and also galactic latipolar
orbits = fit.getOrbit()
orbits_backward = fit_backward.getOrbit()
orbits_full = np.concatenate((orbits, orbits_backward), axis=0)
R, vR, vT, z, vz, phi = orbits_full.T  
X, Y, Z = bovy_coords.galcencyl_to_XYZ(R, phi, z, Xsun=orbigist.rovo[0], Zsun=orbigist.zo).T
l, b, d = bovy_coords.XYZ_to_lbd(X, Y, Z, degree=True).T
```

and then go ahead and produce some sky plots to see how well the fit actually performed. The fit in question is being done for small cold stellar streams, so hoping for somewhat tight orbits that pass all the datapoints- in the case of from_fit() I keep ending up with something like 

![image](https://user-images.githubusercontent.com/16845198/158037047-5dcb82a6-ea79-4416-8aa7-c771e8ab9140.png)

where the dark points are the data, and the red the orbit integral. It would be odd that no proper orbit exists to actually fit the data (given the same stream has been studied before- this is just different data) so I tried writing a fast and dirty least-squares fitter in 6D and sampled a bunch of possible orbits from the data, and did end up with a somewhat convincing (or convincing enough) fit-

![1_all_orbits_test](https://user-images.githubusercontent.com/16845198/158037447-501362df-0783-4b8d-bf6f-8dcdaee4b030.png)

The data still needs a bit of cleaning but at least it's a close-ish fit that shows that the galpy fitter should have at least a few orbits to pick from that somewhat resemble the data (you'll have to take my word for it that the last-squares fit works better than the one galpy produced in all the 6C2 combinations of R/vT/etc...) 

**My question here is... am I interpreting the output of getOrbit() incorrectly?** I'm taking the example from [here](https://docs.galpy.org/en/v1.7.1/orbit.html) for interpreting this. If the fit were correct, this would be the only way I might be handling it poorly, but that doesn't seem to be the case... but you never know? 

**_If_ I'm not wrong in my interpretation of the output of getOrbit(), why is orbit.Orbit.from_fit() not producing a reasonable fit to the data?** As far as I can tell I've followed the instructions for the inputs correctly (with [long,lat,d,mu_ll, mu_bb,vlos] with mu_ll multiplied by cos of the galactic latitude)... for all the stream examples I provide it, the results are similarly off... I doubt the fitter is wrong- if a dirty least-squares one works somewhat, then this one (many times more robust/well-written) should certainly do the trick... am I perhaps misunderstanding the purpose of orbit.from_fit()? 

Thank you and sorry for any trouble this causes, to anyone who replies!

**Comment by jobovy** [comment]
Hi, sorry you're having trouble with this! I don't think there's great documentation for how to use this method, so we should probably improve that!

What you're doing mostly looks right to me, but here are a few comments / things you could try:
* I don't think it works very well to set the error in l and b to zero (even if they are essentially zero). Really, they should be set to the stream width, not the raw uncertainty, because you don't expect the sky positions to match better than the stream width. So you can try setting them to the stream width (probably a few tens of arcmin?). This might be throwing off the fit.
* Are you sure that your mu_ll isn't already multiplied by cos(b)? Catalogs will generally report l (or ra) proper motions with the cos(b) [cos(dec)] factor included, because this is the physical angular velocity (without the cos factor, it's a meaningless coordinate derivative). If you are getting the proper motions from a catalog, they probably have the cos factor already (e.g., Gaia).
* Rather than using `getOrbit` and converting things yourself (potentially doing it wrong), you can just ask for `orbits.ll`, `orbits.bb`, etc (and then concatenate them for leading and trailing)
* Have you tried plotting your initial guess? It's possible it's far enough off that it makes it difficult to find the right solution for the fitter.

I hope this helps! Without having the data directly, it will be difficult to troubleshoot further, but to go further, you could send a subset or make a mock data set from an Orbit (not a bad idea as a test anyway!).

**Comment by callous4567** [comment]
Thanks for the prompt reply! I've followed your advice and can feed back on what worked out well- in order with your bulleted list,


- This solved the problem! (it makes sense in hindsight now that I think about how the fitter may well be trying to force an orbit across a parameter space that may well not have what may appeal to the eye as a "best fit" for the orbit... I expect someone more versed in this field would most likely be aware of this, though I've learned now ^.^) 
- My supervisor actually explained this to me about 2-3 weeks after the MSc started! I kept wondering why all my astropy conversions were so funny- we never got taught this convention... I lost a lot of sleep over it. The galpy documentation staunchly mentions this in many bits though, thankfully, and actually made me aware of it when I glossed over it earlier on in the project before I asked my supervisor.  
- I wrote up a bunch of numba-fuelled coordinate conversions which I spent a lot of time getting to match galpy's that work fairly fast with numba (which I use for the least-squares thing) but I did try this at the start while furiously mulling over what I was doing wrong. 
- The guess wasn't that great compared to the final orbit that galpy provides- I now feed in a weighted least-squares guess, which is much closer (though with the errors added for l/b, this seems to be less of a concern.) 


![image](https://user-images.githubusercontent.com/16845198/158490734-7b7ae1cb-6d44-4bd0-a3f9-a166d825ce4f.png)


Your fitter gave a darn crisp fit! 😁 ... thanks again! My only regret here is not asking earlier- I spent a lot of time trying to figure out what was wrong... more than I'm proud to admit to 😅


The problem is solved but here's the data subset for anyone interested (data in form of HDF5 file... it's a MC sample so not the original data.) 
[bovy.zip](https://github.com/jobovy/galpy/files/8257898/bovy.zip)
